File: backend/code_java/readwordexample/readwordcode/src/main/java/com/code/countword/Implementation/ReadDirectoryImp.java
package com.code.countword.Implementation;

import com.code.countword.util.InfoWordCount;
import com.code.factory.SubscriberFactory;
import com.code.factory.SubscriberFactoryImp;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.RecursiveTask;

public class ReadDirectoryImp extends RecursiveTask<Long>{
    private final InfoWordCount infoWordCount;
    private final SubscriberFactory factory = SubscriberFactoryImp.getInstance();
    public ReadDirectoryImp(InfoWordCount infoWordCount) {
        this.infoWordCount = infoWordCount;
    }

    private Long readDirectoryRecursive(File folder){
        long counter=1L;
        List<RecursiveTask<Long>> task = new ArrayList<>();
       /*
        File[] files = folder.listFiles();
        if(files!=null ) {
            Arrays.stream(files).forEach(t->{
                if(t.isDirectory() && t.canRead())
                {
                    readDirectoryRecursive(t);

                }else{
                    if(t.getAbsolutePath().endsWith(".txt"))
                    {
                        System.out.println(t.getName());
                        ReadFileImp fileImp = factory.getReadFile(t,this.infoWordCount);
                        task.add(fileImp);
                        fileImp.fork();
                        for (RecursiveTask<Long> tas : task) {
                            tas.join();
                        }
                    }
                }});

            return counter;
        }else if(folder.isFile() && folder.getAbsolutePath().endsWith(".txt")) {
            ReadFileImp fileImp = factory.getReadFile(folder,this.infoWordCount);
            task.add(fileImp);
            fileImp.fork();
            for (RecursiveTask<Long> tas : task) {
                tas.join();
            }
            return counter;
        }else {
            return counter;
        }*/
        return Optional.of(folder).map(File::listFiles).map(files->{
            Arrays.stream(files).forEach(file->{
                if(file.isDirectory() && file.canRead()) {
                    readDirectoryRecursive(file);

                }else if(file.getAbsolutePath().endsWith(".txt")) {
                    taskAndMethod(task,factory.getReadFile(file,this.infoWordCount));
                }
            });
            return Optional.of(counter);
        }).orElseGet(()->{
            if(folder.isFile()){
                taskAndMethod(task,factory.getReadFile(folder,this.infoWordCount));
                return Optional.of(counter);
            }
            return Optional.empty();
        }).orElse(counter);

    }
    private void taskAndMethod(List<RecursiveTask<Long>> task, ReadFileImp readFileImp){
        task.add(readFileImp);
        readFileImp.fork();
        task.forEach(RecursiveTask::join);
    }
    @Override
    protected Long compute() {
        return readDirectoryRecursive(new File("C:\\"));
    }
}

File: backend/code_scala/readwordexample/readwordcode/src/main/scala/com/code/countword/implementation/ReadDirectoryImp.scala
package com.code.countword.implementation

import com.code.countword.contract.ReadDirectory
import com.code.countword.util.{AbstractExecutionContext, InfoWordCount}
import com.code.factory.SubscriberFactory

import java.io.File
import scala.annotation.tailrec
import scala.concurrent.Future
import scala.language.postfixOps

class ReadDirectoryImp(infoWordCount: InfoWordCount) extends AbstractExecutionContext with ReadDirectory{
  private val factory = SubscriberFactory
  override def readDirectory(): Future[Option[Long]] = readDirectory(new File("C:\\"))

  override def readDirectory(folder:File): Future[Option[Long]] = {

    @tailrec
    def _readDirectory(files: List[File], listFuture:List[Future[Option[Long]]]=List.empty, count: Int=1):Future[List[Future[Option[Long]]]]= files match {
      case head::tail if head.getAbsolutePath.endsWith(".txt")=>
        println(s"case 1 count: $count name file ${head.getName} num-thread: ${java.lang.Thread.activeCount()}")
        val future = factory.getReadFile(head, infoWordCount).readFile()
        _readDirectory(tail, listFuture:+future, count+1)
      case head::tail if head.isDirectory && Option(head.listFiles).nonEmpty=>
        //println(s"case 2 count: $count name file ${head.getName} num-thread: ${java.lang.Thread.activeCount()}")
        val e = tail ::: head.listFiles.toList
        _readDirectory(tail ::: head.listFiles.toList, listFuture, count+1)
      case head::tail if head.isDirectory && Option(head.listFiles).isEmpty=>
        //println(s"case 3  count: $count name file ${head.getName} num-thread: ${java.lang.Thread.activeCount()}")
        _readDirectory(tail, listFuture, count+1)
      case head::_ if head.isDirectory && Option(head.listFiles).nonEmpty=>
        //println(s"case 4 count: $count name file ${head.getName} num-thread: ${java.lang.Thread.activeCount()}")
        _readDirectory(head.listFiles toList, listFuture, count+1)
      case head::_ if head.getAbsolutePath.endsWith(".txt")=>
        //println(s"case 5 count: $count name file ${head.getName} num-thread: ${java.lang.Thread.activeCount()}")
        val future = factory.getReadFile(head,infoWordCount).readFile()
        Future.successful(listFuture:+future)
      case head::tail if !head.getAbsolutePath.endsWith(".txt")=>
        //println(s"case 6 count: $count name file ${head.getName} num-thread: ${java.lang.Thread.activeCount()}")
         _readDirectory(tail, listFuture, count+1)
      case _=>
        println(s"case final count: $count num-thread: ${java.lang.Thread.activeCount()}")
        Future.successful(listFuture)
    }
   _readDirectory(folder.listFiles toList).map(x=>Future.sequence(x)).flatten
      .flatMap(value=> Future.successful(Option(value.flatten.sum)))
  }
}
